include/fmt/printf.h

```diff
@@ -465,21 +470,21 @@ unsigned basic_printf_context<OutputIt, Char, AF>::parse_header(
 }
 
 template <typename OutputIt, typename Char, typename AF>
-void basic_printf_context<OutputIt, Char, AF>::format() {
-  auto &buffer = internal::get_container(this->out());
+OutputIt basic_printf_context<OutputIt, Char, AF>::format() {
+  auto out = this->out();
   const auto range = this->parse_context();
-  const Char * end = range.end();
-  const Char * start = range.begin();
-  auto it = start;
-  while (it != end) {
+  const Char * const end = range.end();
+  for (const Char *it = range.begin(); it != end; ) {
     char_type c = *it++;
-    if (c != '%') continue;
+    if (c != '%') {
+      *(out++) = c;
+      continue;
+    }
     if (it != end && *it == c) {
-      buffer.append(start, it);
-      start = ++it;
+      ++it;
+      *(out++) = c;
       continue;
     }
-    buffer.append(start, it - 1);
```

Comment:
I think it's better to keep the old logic of writing ranges of characters rather than character by character. Just replace `buffer.append` with `std::copy`. This will be easier to optimize so that reallocations are minimized.